src/BinSkim.Rules/DwarfRules/BA3003.EnableStackProtector.cs

```diff
@@ -38,104 +28,136 @@ public class EnableStackProtector : DwarfSkimmerBase
+            nameof(RuleResources.BA3003_Error),
+            nameof(RuleResources.NotApplicable_InvalidMetadata)
+        };
 
         public override AnalysisApplicability CanAnalyzeDwarf(IDwarfBinary target, Sarif.PropertiesDictionary policy, out string reasonForNotAnalyzing)
         {
-            reasonForNotAnalyzing = null;
+            CanAnalyzeDwarfResult result = default;
+
             if (target is ElfBinary elf)
             {
-                if (elf.ELF.Type == FileType.Executable || elf.ELF.Type == FileType.SharedObject)
-                {
-                    return AnalysisApplicability.ApplicableToSpecifiedTarget;
-                }
-
-                reasonForNotAnalyzing = MetadataConditions.ElfIsCoreNoneOrObject;
+                result = this.VerifyDwarfBinary(elf);
             }
-            else if (target is MachOBinary mainMachO)
+            else if (target is MachOBinary mainMacho)
             {
-                foreach (SingleMachOBinary singleMachO in mainMachO.MachOs)
+                foreach (SingleMachOBinary subMachO in mainMacho.MachOs)
                 {
-                    if (IsApplicableMachO(singleMachO))
+                    result = this.VerifyDwarfBinary(subMachO);
+                    if (result.Result == AnalysisApplicability.ApplicableToSpecifiedTarget)
                     {
-                        // if any macho is applicable
-                        return AnalysisApplicability.ApplicableToSpecifiedTarget;
+                        // if any machO is applicable
+                        break;
                     }
                 }
-
-                reasonForNotAnalyzing = MetadataConditions.MachOIsNotExecutableDynamicLibraryOrObject;
             }
 
-            return AnalysisApplicability.NotApplicableToSpecifiedTarget;
+            reasonForNotAnalyzing = result.Reason;
+            return result.Result;
         }
 
         public override void Analyze(BinaryAnalyzerContext context)
         {
-            HashSet<string> symbolNames = null;
-            bool result = false;
-            if (context.IsELF())
+            IDwarfBinary binary = context.DwarfBinary();
+
+            static bool analyze(IDwarfBinary binary)
             {
-                IELF elf = context.ElfBinary().ELF;
-                symbolNames = new HashSet<string>
+                return binary.CommandLineInfos.All(i =>
                 (
-                    ElfUtility.GetAllSymbols(elf).Select(sym => sym.Name)
-                );
-
-                result = symbolNames.Any(symbol => this.stack_check_symbols.Contains(symbol));
+                    i.CommandLine.Contains("-fstack-protector-all", StringComparison.OrdinalIgnoreCase)
+                    || i.CommandLine.Contains("-fstack-protector-strong", StringComparison.OrdinalIgnoreCase)
+                )
+                && !i.CommandLine.Contains("-fno-stack-protector", StringComparison.OrdinalIgnoreCase));
```

Comment:
Users can control how/when GCC inserts stack protectors with `--param=ssp-buffer-size=N`. If N is set to a very large number, then effectively stack protectors are not used at all. Maybe we should verify that this parameter is not set as well...

Reply:
thanks, will check

Reply:
added the check.